src/infiniteLoops/runtime.ts

```diff
@@ -0,0 +1,295 @@
+  return ignoreMe
+}
+
+/**
+ * Always executed after a loop terminates, or breaks, to clean up
+ * variables and pop the last iteration tracker.
+ */
+function exitLoop(state: st.State) {
+  state.cleanUpVariables()
+  state.exitLoop()
+}
+
+/**
+ * If the number of iterations (given by the length
+ * of stackPositions) is equal to a power of 2 times
+ * the threshold, check these iterations for infinite loop.
+ */
+function checkForInfiniteLoopIfMeetsThreshold(
+  stackPositions: number[],
+  state: st.State,
+  functionName?: string
+) {
+  let checkpoint = state.threshold
+  while (checkpoint <= stackPositions.length) {
+    if (stackPositions.length === checkpoint) {
+      checkForInfiniteLoop(stackPositions, state, functionName)
+    }
+    checkpoint = checkpoint * 2
+  }
+}
+
+/**
+ * Test if stream is infinite. May destructively change the program
+ * environment. If it is not infinite, throw a timeout error.
+ */
+function testIfInfiniteStream(stream: any, state: st.State) {
+  let next = stream
+  for (let i = 0; i <= state.threshold; i++) {
+    if (stdList.is_null(next)) {
+      break
+    } else {
+      const nextTail = stdList.is_pair(next) ? next[1] : undefined
+      if (typeof nextTail === 'function') {
+        next = sym.shallowConcretize(nextTail())
+      } else {
+        break
+      }
+    }
+  }
+  throw new Error('timeout')
+}
+
+const builtinSpecialCases = {
+  is_null(maybeHybrid: any, state?: st.State) {
+    const xs = sym.shallowConcretize(maybeHybrid)
+    const conc = stdList.is_null(xs)
+    const theTail = stdList.is_pair(xs) ? xs[1] : undefined
+    const isStream = typeof theTail === 'function'
+    if (state && isStream) {
+      const lastFunction = state.getLastFunctionName()
+      if (state.streamMode === true && state.streamLastFunction === lastFunction) {
+        // heuristic to make sure we are at the same is_null call
+        testIfInfiniteStream(sym.shallowConcretize(theTail()), state)
+      } else {
+        let count = state.streamCounts.get(lastFunction)
+        if (count === undefined) {
+          count = 1
+        }
+        if (count > state.streamThreshold) {
+          state.streamMode = true
+          state.streamLastFunction = lastFunction
+        }
+        state.streamCounts.set(lastFunction, count + 1)
+      }
+    } else {
+      return conc
+    }
+    return
+  },
+  display: nothingFunction,
+  display_list: nothingFunction
+}
+
+function prepareBuiltins(oldBuiltins: Map<string, any>) {
+  const newBuiltins = new Map<string, any>()
+  for (const [name, fun] of oldBuiltins) {
+    const specialCase = builtinSpecialCases[name]
+    if (specialCase !== undefined) {
+      newBuiltins.set(name, specialCase)
+    } else {
+      newBuiltins.set(name, (...args: any[]) => fun(...args.map(sym.shallowConcretize)))
+    }
+  }
+  newBuiltins.set('undefined', undefined)
+  return newBuiltins
+}
+
+function nothingFunction(...args: any[]) {
+  return nothingFunction
```

Comment:
Why do you return `nothingFunction` instead of, say, `args[0]` or `undefined`? Returning `args[0]` would be more in line with the type of `display`.

Reply:
The `nothingFunction` is used as a stub when some library functions (e.g. from RUNES etc) are used in the code. I didn't want the return value of these functions to be used at all (e.g. `x_of(point) + 1`) and just throw an error when used.